### ISSUE TITLE
Command line build is broken again on Mac

### ISSUE THREAD

**Issue by binarymash** [TomPallister/Ocelot#98]
I noticed the other day that building on Mac is broken. This appears to be because commit efec217c undid some or all of the changes from #41

**Comment by binarymash** [comment]
@TomPallister I assume that this was accidental rather than any intentional change?

**Comment by TomPallister** [comment]
Think this might be a fuck up on my behalf

**Comment by binarymash** [comment]
heh! just about to create a PR... looks like integration tests are failing on the mac but I'll look into that separately.

**Comment by TomPallister** [comment]
@binarymash not sure if this is related but after merging these changes no artifacts are getting pushed to GitHub releases...see this in AppVeyor.

Creating "1.4.5+30.build.214" release for repository "TomPallister/Ocelot" tag "1.4.5+30.build.214" commit "6c1b95f95ba9926cf6f6fb83a5f7c4ce609bb0be"...OK
No artifacts were published. Make sure you have specified correct artifacts filter.

I'm looking into it but thought I would let you know in case you can provide any assistance!

Ta!

**Comment by TomPallister** [comment]
There is a input 

Artifact(s) to deploy

In AppVeyor for the release preparation environment. This doesn't have anything in it. Trying to work out if I'm in the right area!

**Comment by binarymash** [comment]
Sorry, only just saw this. Did you get it working? I'll need to familiarise myself with the build process again to investigate.

**Comment by TomPallister** [comment]
@binarymash no worries!! Nope I didn't get it working.

**Comment by binarymash** [comment]
Looking in appveyor the first build to not have artifacts seems to be 1.4.5-PullRequest.104+7.build.202 (see https://ci.appveyor.com/project/TomPallister/ocelot-fcfpb/build/1.4.5-PullRequest.104+7.build.202)

**Comment by binarymash** [comment]
Oh, hang on, looking into the actual build logs, we're generally not seeing artifacts because the build is failing in the CreatePackages step. Not sure why the build shows up as successful. Looks like it's something to do with generating the release notes.

```
========================================
CreatePackages
========================================
Executing task: CreatePackages
Creating directory C:/projects/ocelot-fcfpb/artifacts/Packages
Copying file Ocelot.1.5.0-unstable0102.nupkg to C:/projects/ocelot-fcfpb/artifacts/Packages/Ocelot.1.5.0-unstable0102.nupkg
Copying file Ocelot.1.5.0-unstable0102.nupkg to C:/projects/ocelot-fcfpb/artifacts/Packages/Ocelot.1.5.0-unstable0102.nupkg
Generating release notes at artifacts/Packages/releasenotes.md
Tag refs/tags/1.2.0-unstable not added to the release list, because a tag for that commit was added already.
Tag refs/tags/1.3.0-unstable not added to the release list, because a tag for that commit was added already.
Cake.exe : An error occurred when executing task 'CreatePackages'.
At line:1 char:1
+ & "C:\projects\ocelot-fcfpb\tools\Cake\Cake.exe" "build.cake" -target ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (An error occurr...reatePackages'.:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
Error: 
Could not find file 'C:\projects\ocelot-fcfpb\artifacts\Packages\releasenotes.md'.
Build success
```

**Comment by TomPallister** [comment]
ive removed the release notes publishing for now, trying to get a build out without them!

**Comment by TomPallister** [comment]
@binarymash i removed the releasenotes code and now get an error when trying to push the release artifacts from AppVeyor to GitHub.

Creating "1.4.5+33.build.222" release for repository "TomPallister/Ocelot" tag "1.4.5+33.build.222" commit "6786b5b0afa4fc6b9ad1253e76f1e8ad02f19703"...Skipped - release with tag "1.4.5+33.build.222" already exists
Uploading "artifacts.txt" to release assets...An error occurred while sending the request.

No idea why this is happening. I tried reverting the build for mac commit but this didn't help (ive not put that commit back in). I also tried generating a new GitHub token for the release environment. This didn't make a difference either.

Any ideas? I haven't changed the release prep settings as far as Im aware and the last time it worked was 2 months ago :(

**Comment by TomPallister** [comment]
@binarymash also created this on AppVeyor forums...

http://help.appveyor.com/discussions/problems/6892-cant-publish-artifacts-to-github-release

**Comment by binarymash** [comment]
Gonna have a quick look at this now. Does it generate the release notes when you build locally?

**Comment by binarymash** [comment]
Looks like you've got some strange branching going on. I noticed that build 1.4.5+3.build.200 (sha 89375), which is on master, appears to be a merge of a pull request sha. This build also fails with the release notes failing. The whole build/release process assumes we're using gitflow branching, so merging from PR to master isn't allowed - should be merged to develop, and then a release branch cut before merging to master. No idea if this is messing up the versioning in some way.

The versioning also looks a bit off and I can't understand what's going on... I see things like the commit on release/1.0.6 is tagged with 1.4.3, the commit on release/1.0.7 is tagged with 1.4.4, and so on. I don't understand this... gitversion uses tags and release branch names to decide what numbers to use for builds, so I would expect to see that after release/1.0.6 the master is tagged with 1.0.6. Take a look at the tags/branches on https://github.com/binarymash/responses for how it should look. How have you been doing releases/tags on master? Where's that document I wrote about the build process - can't find it now.

So, I think maybe the combination of weird versioning and dodgy branching means that gitversion/gitreleasenotes, which rely on certain conventions, are getting confused. If that's the case I'm not quite sure what the solution is.. think there might be an option to tell gitversion to ignore commits before a certain sha, which might let us get back on track with that.

**Comment by binarymash** [comment]
Looks like you already have a draft release for 1.4.5+33.build.222 in github - see https://github.com/TomPallister/Ocelot/releases . If you're trying to deploy this again from AppVeyor then you'll need to delete the draft first.

**Comment by TomPallister** [comment]
@binarymash turns out its an AppVeyor/GitHub issue in terms of not being able to publish artifacts.

AppVeyor have raised it with GitHub. Thanks for looking at it for me...feel like I wasted your time.

I thought I was doing my branching correctly. I have pushing my commits to develop and then creating a release branch...push that to remote. Then I merge that release branch into master and that's the build I deploy to GitHub. Seems to be working so far. Ignore those release branches numbering I didn't think it mattered and was just an arbitrary number that popped into my head. I always delete them.

The general take away is that I've messed up the GitVersion stuff and that's why I'm not getting release notes...will follow this avenue!

**Comment by TomPallister** [comment]
@binarymash I do get release notes when I build locally.

**Comment by binarymash** [comment]
ah ok no worries. 

Yeah sounds like your branching is generally correct - merges to develop and master, commits to features and release. I'd definitely create the release branches with the version number of what you're actually going to release, as it's pretty confusing otherwise. I think once you've created the release branch this would also potentially affect the versioning on subsequent commits to develop, before you even do the final merge to master/develop - for example, if you bump the major version on the release branch from 1.x to 2.x, subsequent commits on develop will be automatically versioned as 2.x even before you merge to master.Also means that the release created in github automatically has the correct number. Basically the only decision to really make about versioning is on the release branch... once you've chosen it here, everything else should just flow automatically.

I actually always do a commit on the release branch, to copy the contents of the generated release notes into a releasenotes.md (I also do a bit of manual editing to add in a date and format it) - see https://github.com/binarymash/responses/blob/develop/releasenotes.md - gives a handy reference of all the changes over time. Another reason to use the correct version number on the release branch.

Anyway, good to hear it's nothing wrong on our side. Fingers crossed they get it sorted.

**Comment by TomPallister** [comment]
@TomPallister - thanks!. I just pulled down your changes and gave it a try. Same error. Note, I am on a mac os.

Output from running ./build.sh is below. The only way I am able to remedy the problem is the workaround I described above. I also pulled down Cake.net's example template and tried the .ToDictionary() extension and received the same error therefore the issue is unrelated to Ocelot. However, the build does not seem to run on macos - I wonder if other non-windows users are experiencing the same issue. I just posted the error on the Cake-build Gitter channel, hopefully, someone will provide insight.

Output from running ./build.sh:

Analyzing build script...
Processing build script...
Installing tools...
Installing addins...
Compiling build script...
Error: Error occurred when compiling build script: /Users/stuart/Dev/src/dotnet/Ocelot/build.cake(378,25): error CS0012: The type 'Dictionary<,>' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Collections, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.

**Comment by stuartsaltzman** [comment]
@TomPallister - The newly released Cake version 0.25.0 resolves the issue experienced when building on mac (mono) as described above. I tried building on mac using Cake v0.25.0 and it builds successfully without the workaround.

Cake 0.25.0 release notes (https://www.nuget.org/packages/Cake/0.25.0):
- ToDictionary on Mono causes "The type 'Dictionary<,>' is defined in an assembly that is not referenced"
